### PR TITLE
Remove friendship classes from Graph and replace with Unsafe functions instead.

### PR DESCRIPTION
**Pull request by Relux-the-Relux**
Addressing #740.

### CODE REVIEW

File: include/networkit/graph/Graph.hpp [hunk 1]
```diff
@@ -1441,6 +1633,15 @@ class Graph final {
...
     void increaseWeight(node u, node v, edgeweight ew);
 
+    /**
+     * Increase the weight of i-th edge in the adjecency array of u.
+     *
+     * @param[in]	u	endpoint of edge
+     * @param[in]	i	index of edge in the adjacency array for u.
+     * @param[in]	weight	edge weight
+     */
+    void increaseWeightAtItNeighbour(Unsafe, node u, index i, edgeweight ew);
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply include/networkit/graph/Graph.hpp:1643]: oh yeah

File: include/networkit/graph/Graph.hpp [hunk 2]
```diff
@@ -1362,6 +1523,37 @@ class Graph final {
      */
     count numberOfNodes() const noexcept { return n; }
 
+    /**
+     * Returns the size of the outEdges vector. This most is likely the number of nodes but might
+     * differ if nodes have been deleted.
+     * @return The size of outEdges.
+     */
+    count numberOfNodes(Unsafe) const noexcept { return outEdges.size(); }
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply include/networkit/graph/Graph.hpp:1531]: the problem with `numberOfWeightedNodes` and `numberOfWeightedInNodes` needs to be 0 in the unweighted/undirected when `GraphBuilder` checks for the consistency of the graph.

maybe we could simply make these private and add the checks to `checkConsistency()` and remove the assertions from `GraphBuilder` replacing them only with the `checkConsistency()` call.

File: networkit/cpp/randomization/CurveballImpl.cpp [hunk 3]
```diff
@@ -154,29 +154,35 @@ void CurveballMaterialization::toGraphParallel(Graph &G) {
...
+    for (index u = 0; u < new_outEdges.size(); u++) {
+        for (index j = 0; j < new_outEdges[u].size(); j++) {
+            G.addPartialEdge(unsafe, u, new_outEdges[u][j]);
+        }
+    }
 
     // Reserve the space
-    G.parallelForNodes([&](node v) { G.outEdges[v].reserve(outDeg[v] + missingEdgesCounts[v]); });
+    G.parallelForNodes(
+        [&](node v) { G.preallocateUndirected(v, outDeg[v] + missingEdgesCounts[v]); });
```
Comment (avdgrinten) [inline networkit/cpp/randomization/CurveballImpl.cpp:166]: The `addPartialEdge` loop can be moved into the parallelForNodes below (after preallocation).

File: networkit/cpp/randomization/CurveballImpl.cpp [hunk 4]
```diff
@@ -154,29 +154,35 @@ void CurveballMaterialization::toGraphParallel(Graph &G) {
...
+            const node u = G.getIthNeighbor(v, neighbor_id);
+            G.addPartialEdge(unsafe, u, v);
         }
     });
 
     // TODO: is the networkit adjacency list even sorted for the neighbours? if
     // not omit this
     // Sort neighbours
-    G.parallelForNodes([&](node v) { std::sort(G.outEdges[v].begin(), G.outEdges[v].end()); });
+    // G.parallelForNodes([&](node v) { std::sort(G.outEdges[v].begin(), G.outEdges[v].end()); });
```
Comment (avdgrinten) [inline networkit/cpp/randomization/CurveballImpl.cpp:179]: Remove this chunk (as the comment states).

File: networkit/cpp/randomization/CurveballImpl.cpp [hunk 5]
```diff
@@ -206,29 +212,34 @@ void CurveballMaterialization::toGraphSequential(Graph &G) {
...
+    G.removeAllEdges();
+    for (index u = 0; u < new_outEdges.size(); u++) {
+        for (index j = 0; j < new_outEdges[u].size(); j++) {
+            G.addPartialEdge(unsafe, u, new_outEdges[u][j]);
+        }
+    }
 
     // Reserve the space
-    G.forNodes([&](node v) { G.outEdges[v].reserve(outDeg[v] + missingEdgesCounts[v]); });
+    G.forNodes([&](node v) { G.preallocateUndirected(v, outDeg[v] + missingEdgesCounts[v]); });
```
Comment (avdgrinten) [inline networkit/cpp/randomization/CurveballImpl.cpp:223]: See above.

File: networkit/cpp/randomization/CurveballImpl.cpp [hunk 6]
```diff
@@ -206,29 +212,34 @@ void CurveballMaterialization::toGraphSequential(Graph &G) {
...
+            const node u = G.getIthNeighbor(v, neighbor_id);
+            G.addPartialEdge(unsafe, u, v);
         }
     });
 
     // TODO: is the networkit adjacency list even sorted for the neighbours? if
     // not omit this
     // Sort neighbours
-    G.forNodes([&](node v) { std::sort(G.outEdges[v].begin(), G.outEdges[v].end()); });
+    // G.forNodes([&](node v) { std::sort(G.outEdges[v].begin(), G.outEdges[v].end()); });
```
Comment (avdgrinten) [inline networkit/cpp/randomization/CurveballImpl.cpp:236]: See above.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 7]
```diff
@@ -145,11 +145,8 @@ void GraphBuilder::increaseInWeight(node u, node v, edgeweight ew) {
 
 Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
     Graph G(n, weighted, directed);
-    assert(G.outEdges.size() == n);
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
+    assert(G.checkConsistency(n));
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:148]: `checkConsistency` is expensive and should be behind `NETWORKIT_SANITY_CHECKS`.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 8]
```diff
@@ -161,12 +158,8 @@ Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
         toGraphDirectSwap(G);
     }
 
-    assert(G.outEdges.size() == n);
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
-    G.m = numberOfEdges(G);
+    assert(G.checkConsistency(n));
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:161]: Likewise.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 9]
```diff
@@ -213,24 +226,26 @@ void GraphBuilder::toGraphParallel(Graph &G) {
                                         // undirected graphs
                 inEdgesPerThread[tid][u].push_back(v);
                 if (weighted) {
-                    edgeweight ew = outEdgeWeights[v][i];
-                    inWeightsPerThread[tid][u].push_back(ew);
+                    G.addPartialEdge(unsafe, v, u, outEdgeWeights[v][i]);
+                    inWeightsPerThread[tid][u].push_back(outEdgeWeights[v][i]);
+                } else {
+                    G.addPartialEdge(unsafe, v, u);
                 }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:233]: Is adding edges safe here? I assume no two threads access the same `v` at the same time?

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 10]
```diff
@@ -213,24 +226,26 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
             }
             if (u == v) {
                 numberOfSelfLoopsPerThread[tid]++;
+                if(!directed){
+                    inEdgesPerThread[tid][u].push_back(v);
+                    if (weighted) inWeightsPerThread[tid][u].push_back(outEdgeWeights[v][i]);
+                }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:240]: Why is this addition necessary?

Comment (avdgrinten) [reply networkit/cpp/graph/GraphBuilder.cpp:240]: The formatting of the `if` is broken.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 11]
```diff
@@ -239,60 +254,75 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
-                G.inEdgeWeights[v].reserve(inDeg);
-            }
-        } else {
-            G.outEdges[v].reserve(outDeg + inDeg);
-            if (weighted) {
-                G.outEdgeWeights[v].reserve(outDeg + inDeg);
-            }
+            G.preallocateDirectedInEdges(v, inDeg);
+            } else {
+            G.preallocateUndirected(v, outDeg + inDeg);
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:259]: Formatting is broken.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 12]
```diff
@@ -239,60 +254,75 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
         if (directed) {
-            for (int tid = 0; tid < maxThreads; tid++) {
-                copyAndClear(inEdgesPerThread[tid][v], G.inEdges[v]);
-            }
             if (weighted) {
                 for (int tid = 0; tid < maxThreads; tid++) {
-                    copyAndClear(inWeightsPerThread[tid][v], G.inEdgeWeights[v]);
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                    G.addPartialInEdge(unsafe, v, inEdgesPerThread[tid][v][j], inWeightsPerThread[tid][v][j]);
+                    }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:268]: Formatting is broken.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 13]
```diff
@@ -239,60 +254,75 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
         } else {
-            for (int tid = 0; tid < maxThreads; tid++) {
-                copyAndClear(inEdgesPerThread[tid][v], G.outEdges[v]);
-            }
             if (weighted) {
                 for (int tid = 0; tid < maxThreads; tid++) {
-                    copyAndClear(inWeightsPerThread[tid][v], G.outEdgeWeights[v]);
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                    G.addPartialEdge(unsafe, v, inEdgesPerThread[tid][v][j], inWeightsPerThread[tid][v][j]);
+                    }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:282]: Formatting is broken.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 14]
```diff
@@ -239,60 +254,75 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
-                    copyAndClear(inWeightsPerThread[tid][v], G.inEdgeWeights[v]);
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                    G.addPartialInEdge(unsafe, v, inEdgesPerThread[tid][v][j], inWeightsPerThread[tid][v][j]);
+                    }
+                }
+            } else {
+                for (int tid = 0; tid < maxThreads; tid++) {
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                        G.addPartialInEdge(unsafe, v, inEdgesPerThread[tid][v][j]);
+                    }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:274]: Why do we need a new else case here?

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 15]
```diff
@@ -239,60 +254,75 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
-                    copyAndClear(inWeightsPerThread[tid][v], G.outEdgeWeights[v]);
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                    G.addPartialEdge(unsafe, v, inEdgesPerThread[tid][v][j], inWeightsPerThread[tid][v][j]);
+                    }
+                }
+            } else {
+                for (int tid = 0; tid < maxThreads; tid++) {
+                    for (index j = 0; j < inEdgesPerThread[tid][v].size(); j++){
+                        G.addPartialEdge(unsafe, v, inEdgesPerThread[tid][v][j]);
+                    }
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:288]: Why do we need a new else case here?

File: include/networkit/graph/Graph.hpp [hunk 16]
```diff
@@ -1441,6 +1459,15 @@ class Graph final {
...
     void increaseWeight(node u, node v, edgeweight ew);
 
+    /**
+     * Increase the weight of i-th edge in the adjecency array of u.
+     *
+     * @param[in]	u	endpoint of edge
+     * @param[in]	i	index of edge in the adjacency array for u.
+     * @param[in]	weight	edge weight
+     */
+    void increaseWeightAtIthNeighbor(Unsafe, node u, index i, edgeweight ew);
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply include/networkit/graph/Graph.hpp:1469]: It was more a question of speed that I implemented as a separate function, to avoid having to call the weight everytime we want to increase it.

File: include/networkit/graph/Graph.hpp [hunk 17]
```diff
@@ -1521,6 +1548,15 @@ class Graph final {
...
     }
 
+    /**
+     * Returns the index of node v in the array of outgoing edges of node u.
+     *
+     * @param u Node
+     * @param v Node
+     * @return index of node v in the array of outgoing edges of node u.
+     */
+    index indexOfNeighbor(Unsafe, node u, node v) const { return indexInOutEdgeArray(u, v); }
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply include/networkit/graph/Graph.hpp:1558]: you're right it shouldn't be

File: networkit/cpp/graph/GraphBuilder.cpp (hunk 14 above)
Comment (Relux-the-Relux) [reply networkit/cpp/graph/GraphBuilder.cpp:274]: It is so we don't try to access the weight array when we don't have any weights, which would cause a segmentation fault

File: include/networkit/graph/Graph.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (avdgrinten) [reply include/networkit/graph/Graph.hpp:1469]: But aside from safety checks (which are omitted by the `Unsafe` variant anyway), that is equivalent, right? Surely the compiler is able to inline the call (at least, if we define the function in the header) and will generate identical assembly?

File: networkit/cpp/graph/GraphBuilder.cpp (hunk 14 above)
Comment (avdgrinten) [reply networkit/cpp/graph/GraphBuilder.cpp:274]: Makes sense. How did the old code deal with that?

Comment (Relux-the-Relux) [reply networkit/cpp/graph/GraphBuilder.cpp:274]: it just replaced both arrays in graph, since it would just put an array of size 0 into in the unweighted case.

File: include/networkit/graph/Graph.hpp [hunk 18]
```diff
@@ -1535,6 +1571,36 @@ class Graph final {
...
+     * @return pair: i-th (outgoing) neighbor of @a u and the corresponding
+     * edge weight, or @c defaultEdgeWeight if unweighted.
+     */
+    std::pair<node, edgeweight> getIthNeighborWithWeight(node u, index i) const {
+        if (!hasNode(u) || i >= outEdges[u].size())
+            return {none, none};
+        if (!isWeighted())
+            return {outEdges[u][i], defaultEdgeWeight};
+        return {outEdges[u][i], outEdgeWeights[u][i]};
+    }
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply include/networkit/graph/Graph.hpp:1588]: Yes it is used in graphbuilder

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 19]
```diff
@@ -161,12 +158,8 @@ Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
...
     }
 
-    assert(G.outEdges.size() == n);
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
-    G.m = numberOfEdges(G);
+    assert(G.checkConsistency(n));
+    G.setEdgeCount(unsafe, numberOfEdges(G));
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply networkit/cpp/graph/GraphBuilder.cpp:162]: Because it only checks for number for the node upper bound, the number of edges is not relevant for cosistency.

File: networkit/cpp/graph/GraphBuilder.cpp (hunk 9 above)
Comment (Relux-the-Relux) [reply networkit/cpp/graph/GraphBuilder.cpp:233]: yes it is safe

File: include/networkit/graph/Graph.hpp [hunk 20]
```diff
@@ -88,9 +88,9 @@ class CurveballMaterialization;
  */
 class Graph final {
 
-    friend class ParallelPartitionCoarsening;
-    friend class GraphBuilder;
-    friend class CurveballDetails::CurveballMaterialization;
+    // friend class ParallelPartitionCoarsening;
+    // friend class GraphBuilder;
+    // friend class CurveballDetails::CurveballMaterialization;
```
Comment (avdgrinten) [inline include/networkit/graph/Graph.hpp:93]: This needs to be fixed.

File: include/networkit/graph/Graph.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (avdgrinten) [reply include/networkit/graph/Graph.hpp:1473]: What's the progress here? I really don't want an "increase" function as a basic building block.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 21]
```diff
@@ -145,11 +145,11 @@ void GraphBuilder::increaseInWeight(node u, node v, edgeweight ew) {
...
     Graph G(n, weighted, directed);
-    assert(G.outEdges.size() == n);
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
+    #ifdef NETWORKIT_SANITY_CHECKS
+    assert(n == G.upperNodeIdBound());
+    assert(G.checkConsistency());
+    #endif
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:151]: The cheap `n == G.upperNodeIdBound()` assertion can stay outside of NETWORKIT_SANITY_CHECKS.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 22]
```diff
@@ -160,13 +160,11 @@ Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
...
-    assert(G.outEdges.size() == n);
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
-    G.m = numberOfEdges(G);
+    #ifdef NETWORKIT_SANITY_CHECKS
+    assert(n == G.upperNodeIdBound());
+    assert(G.checkConsistency());
+    #endif
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:166]: Likewise.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 23]
```diff
@@ -213,24 +231,26 @@ void GraphBuilder::toGraphParallel(Graph &G) {
...
+                    inWeightsPerThread[tid][u].push_back(outEdgeWeights[v][i]);
+                } else {
+                    G.addPartialEdge(unsafe, v, u);
                 }
             }
             if (u == v) {
                 numberOfSelfLoopsPerThread[tid]++;
+                if (!directed) {
+                    inEdgesPerThread[tid][u].push_back(v);
+                    if (weighted) inWeightsPerThread[tid][u].push_back(outEdgeWeights[v][i]);
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:244]: Coding style.

File: networkit/cpp/randomization/CurveballImpl.cpp [hunk 24]
```diff
@@ -206,29 +206,31 @@ void CurveballMaterialization::toGraphSequential(Graph &G) {
         }
     });
 
-    G.outEdges.swap(new_outEdges);
+    G.removeAllEdges();
+    for (index u = 0; u < new_outEdges.size(); u++) {
+    }
```
Comment (avdgrinten) [inline networkit/cpp/randomization/CurveballImpl.cpp:211]: Huh?

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 25]
```diff
@@ -160,13 +160,11 @@ Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
...
-    assert(G.outEdgeWeights.size() == (weighted ? n : 0));
-    assert(G.inEdges.size() == (directed ? n : 0));
-    assert(G.inEdgeWeights.size() == ((weighted && directed) ? n : 0));
-
-    G.m = numberOfEdges(G);
+    #ifdef NETWORKIT_SANITY_CHECKS
+    assert(n == G.upperNodeIdBound());
+    assert(G.checkConsistency());
+    #endif
+    G.setEdgeCount(unsafe, numberOfEdges(G));
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphBuilder.cpp:167]: `checkConsistency` should be adjusted to also check the edge count. (And hence `setEdgeCount()` should be called before `checkConsistency()`.)

**Review by avdgrinten — COMMENTED** [review]
Looks mostly good now. Please do a final pass to resolve the comments and:
* Cleanup the code by removing the commented out parts. There are also `if`s that have the `then-case on the same lines; these should be fixed.
* Tidy up the commit structure, by removing the commits that fix earlier commits. It might make sense to squash everything except for the documentation additions into a single commit.

**Comment by michitux** [comment]
Do you have any idea what impact these changes have on performance and memory usage? From a short look, this should for example duplicate the memory usage of `METISGraphReader` as the graph is now copied from `GraphBuilder` into the graph instead of swapping the edge data structures which is imho not acceptable. Further, the method name `toGraphDirectSwap` is now completely misleading, as it is no longer a direct swap but a fully sequential copy that does not even pre-allocate the vectors. It should be renamed, which is possible as it is a private method. As mentioned above, I would suggest to eliminate all uses of `GraphBuilder` that has `autoCompleteEdges` set to `false`.

File: networkit/cpp/graph/GraphBuilder.cpp [hunk 26]
```diff
@@ -175,14 +173,34 @@ Graph GraphBuilder::toGraph(bool autoCompleteEdges, bool parallel) {
 }
 
 void GraphBuilder::toGraphDirectSwap(Graph &G) {
-    G.outEdges = std::move(outEdges);
-    G.outEdgeWeights = std::move(outEdgeWeights);
-    G.inEdges = std::move(inEdges);
-    G.inEdgeWeights = std::move(inEdgeWeights);
+    G.removeAllEdges();
```
Comment (michitux) [inline networkit/cpp/graph/GraphBuilder.cpp:176]: This is a private method that is only called with an empty graph. An assert that the graph is empty would be more appropriate imho.

Comment (Relux-the-Relux) [reply networkit/cpp/graph/GraphBuilder.cpp:176]: But i don't think the Graph would be empty. DirectSwap we expect the graph to not be empty.

Comment (michitux) [reply networkit/cpp/graph/GraphBuilder.cpp:176]: `toGraphDirectSwap` is only called in `toGraph` (it is private). There, the graph object is initialized with `Graph G(n, weighted, directed);`. I cannot see any code there that could add edges to the graph that is called before `toGraphDirectSwap`.

**Comment by avdgrinten** [comment]
I am aware that this is a performance pessimization. Michael is probably right about rewriting the users of GraphBuilder that do not need edge completion (these can directly use the unsafe methods). GraphBuilder itself can be re-written to use a better algorithm (there are smarter transpose algorithms than what we implement right now). This can probably also half the memory consumption and get it back closer to the original memory consumption (by only storing the half of the edges that still requires completion and directly inserting the other half).

I think that should be done in a separate PR though (to avoid making this PR more complicated). This means that we will not merge this PR before the next release (which is scheduled for June 30th) since it doesn't make sense to merge a performance pessimization right before the release. We should still complete this PR now, immediately merge it after the release and then optimize `GraphBuilder`.

**Review by avdgrinten — APPROVED** [review]
